generator/mavgen.py

```diff
@@ -64,38 +64,82 @@ def mavgen(opts, args):
             print("WARNING: Unable to load XML validator libraries. XML validation will not be performed", file=sys.stderr)
             opts.validate = False
 
+
     def expand_includes():
-        """Expand includes in current list of all files, ignoring those already parsed."""
-        for x in xml[:]:
-            for i in x.include:
-                fname = os.path.join(os.path.dirname(x.filename), i)
-
-                # Only parse new include files
-                if fname in all_files:
-                    continue
-                all_files.add(fname)
-
-                # Validate XML file with XSD file if possible.
-                if opts.validate:
-                    print("Validating %s" % fname)
-                    if not mavgen_validate(fname):
-                        return False
-                else:
-                    print("Validation skipped for %s." % fname)
-
-                # Parsing
-                print("Parsing %s" % fname)
-                xml.append(mavparse.MAVXML(fname, opts.wire_protocol))
-
-                # include message lengths and CRCs too
-                x.message_crcs.update(xml[-1].message_crcs)
-                x.message_lengths.update(xml[-1].message_lengths)
-                x.message_min_lengths.update(xml[-1].message_min_lengths)
-                x.message_flags.update(xml[-1].message_flags)
-                x.message_target_system_ofs.update(xml[-1].message_target_system_ofs)
-                x.message_target_component_ofs.update(xml[-1].message_target_component_ofs)
-                x.message_names.update(xml[-1].message_names)
-                x.largest_payload = max(x.largest_payload, xml[-1].largest_payload)
+        """Expand includes. XML files from arguments already parsed inot xml list."""
+        
+        def expand_oneiteration():
+            noincludeadded = True
+
+            for x in xml[:]:
+                try:
+                    if not x.parent:
+                        pass
+                except:
+                    x.parent=set()
+
+                for i in x.include:
+                    fname = os.path.join(os.path.dirname(x.filename), i)
+                    # Only parse new include files
+                    if fname in all_files:
+                        continue
+                    all_files.add(fname)
+                    # Validate XML file with XSD file if possible.
+                    if opts.validate:
```

Comment:
We're also validating at the top level.  What's up with that?

Reply:
We've validated the top level files already. This is validating (just) the included files as well.